File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph, visited, q = defaultdict(list), set(), deque([source])
        
        for ed in edges:
            x, y = ed[0], ed[1]
            graph[x].append(y)
            graph[y].append(x)
            
        while q:
            node = q.popleft()
            if node == destination: return True
            for n in graph[node]:
                if n not in visited:
                    q.append(n)
                    visited.add(n)
        return False
              
                
#         def dfs(src, dest):
#             visited.add(src)
#             for x in graph[src]:
#                 if x not in visited:
#                     dfs(x, dest)
#             return dest in visited
        
#         return dfs(source, destination)

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        
        vector<vector<int>> adj(n);
        for (auto e: edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        vector<int>vis(n);
        queue<int> q;
        q.push(source);
        
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            if (node == destination)
                return true;
            for (auto u: adj[node]) {
                if (!vis[u])
                    q.push(u);
                vis[u] = true;
            }
        }
        return false;
    }
};